### PR TITLE
Add contrib.compute_accidental_hits operator for candidate sampling

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
reference operator: https://www.tensorflow.org/api_docs/python/tf/nn/compute_accidental_hits 
The operator computes the matching pair ("accidental hits") between true classes and samples. 
```
Compute the indices in ``sampled_candidates`` which matches ``true_classes``
and return the mask for the matching positions.

The operator is used for removing sampled classes which happen to match target classes
(i.e. accidental hits) for sampled softmax and sampled logistic. The mask has 0
for non-matching positions and 1 for matching ones.

Both inputs are expected to be 1-D. For example, let's say ``true_classes`` has shape (M,),
and ``sampled_candidates`` has (N,), then the resulting mask will have shape (M, N).
mask[i][j] = 1 iff true_classes[i] == sampled_candidates[j].

Example::

   true = [1,5,11]
   sampled = [5,8,1,5,24]
   compute_accidental_hits(true, sampled) = [[0, 0, 1, 0, 0],
                                             [1, 0, 0, 1, 0],
                                             [0, 0, 0, 0, 0]]

.. note:: `compute_accidental_hits` is only available on CPU and returns a compressed sparse row mask.
```

The main difference between this operator and the original tf operator is, the tf operator directly returns the indices of matching positions, which doesn't have a fixed shape. This operator instead returns a mask with all matching positions set to 1. 

The resulting mask can be used with `nd.where` to set the values for the accidental hits. 

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: tests/python/unittest/test_sparse_ndarray.py [hunk 1]
```diff
@@ -843,6 +843,42 @@ def check_sparse_nd_norm(stype, shape, density):
...
+        accidental_hits = []
+        for i in range(num_sample):
+            s = sample[i]
+            if s in label_indices:
+                hits = label_indices[s]
+                accidental_hits += [(h, i) for h in hits]
+        return accidental_hits
+
+    n = 20
+    num_label = np.random.randint(1, 10)
```
Comment (marcoabreu) [inline tests/python/unittest/test_sparse_ndarray.py]: Please initialize a fixed seed

Comment (eric-haibin-lin) [reply tests/python/unittest/test_sparse_ndarray.py]: added

**Comment by eric-haibin-lin** [comment]
Requires https://github.com/dmlc/mshadow/pull/324

**Comment by sxjscience** [comment]
Is the functionality something like `broadcast_equal`?

**Comment by sxjscience** [comment]
I see, it returns a CSR matrix.

**Comment by sxjscience** [comment]
Looks good once the problem about hashing is fixed.

**Comment by eric-haibin-lin** [comment]
Closing it for now until the hash issue is fixed.

**Pull request closed without merging**